### PR TITLE
[APPSRE-2611] Tracking decorator

### PR DESCRIPTION
**Pull request by apahim**
```
qontract_reconcile_function_elapsed_seconds_since_bundle_commit_bucket{integration="openshift-resources",le="10.0",name="utils.oc.OC.apply"} 0.0
qontract_reconcile_function_elapsed_seconds_since_bundle_commit_bucket{integration="openshift-resources",le="30.0",name="utils.oc.OC.apply"} 0.0
qontract_reconcile_function_elapsed_seconds_since_bundle_commit_bucket{integration="openshift-resources",le="60.0",name="utils.oc.OC.apply"} 0.0
qontract_reconcile_function_elapsed_seconds_since_bundle_commit_bucket{integration="openshift-resources",le="150.0",name="utils.oc.OC.apply"} 0.0
qontract_reconcile_function_elapsed_seconds_since_bundle_commit_bucket{integration="openshift-resources",le="300.0",name="utils.oc.OC.apply"} 10.0
qontract_reconcile_function_elapsed_seconds_since_bundle_commit_bucket{integration="openshift-resources",le="600.0",name="utils.oc.OC.apply"} 0.0
qontract_reconcile_function_elapsed_seconds_since_bundle_commit_bucket{integration="openshift-resources",le="1200.0",name="utils.oc.OC.apply"} 0.0
qontract_reconcile_function_elapsed_seconds_since_bundle_commit_bucket{integration="openshift-resources",le="1800.0",name="utils.oc.OC.apply"} 0.0
qontract_reconcile_function_elapsed_seconds_since_bundle_commit_bucket{integration="openshift-resources",le="+Inf",name="utils.oc.OC.apply"} 0.0
...
qontract_reconcile_function_elapsed_seconds_since_bundle_commit_bucket{integration="openshift-resources",le="10.0",name="utils.oc.OC.delete"} 0.0
qontract_reconcile_function_elapsed_seconds_since_bundle_commit_bucket{integration="openshift-resources",le="30.0",name="utils.oc.OC.delete"} 0.0
qontract_reconcile_function_elapsed_seconds_since_bundle_commit_bucket{integration="openshift-resources",le="60.0",name="utils.oc.OC.delete"} 0.0
qontract_reconcile_function_elapsed_seconds_since_bundle_commit_bucket{integration="openshift-resources",le="150.0",name="utils.oc.OC.delete"} 2.0
qontract_reconcile_function_elapsed_seconds_since_bundle_commit_bucket{integration="openshift-resources",le="300.0",name="utils.oc.OC.delete"} 0.0
qontract_reconcile_function_elapsed_seconds_since_bundle_commit_bucket{integration="openshift-resources",le="600.0",name="utils.oc.OC.delete"} 0.0
qontract_reconcile_function_elapsed_seconds_since_bundle_commit_bucket{integration="openshift-resources",le="1200.0",name="utils.oc.OC.delete"} 0.0
qontract_reconcile_function_elapsed_seconds_since_bundle_commit_bucket{integration="openshift-resources",le="1800.0",name="utils.oc.OC.delete"} 0.0
qontract_reconcile_function_elapsed_seconds_since_bundle_commit_bucket{integration="openshift-resources",le="+Inf",name="utils.oc.OC.delete"} 0.0
```

### CODE REVIEW

**Comment by mmclanerh** [comment]
super cool! :+1:

**Comment by rporres** [comment]
While this is certainly interesting and I love the idea of being able to measure the time that integrations take to run, I fail to understand how it help us to measure the time from merge to apply... Is this part of larger plan?

**Comment by jmelis** [comment]
I'm concerned this may cause a cardinality explosion. It basically implies ~20 new time-series per git commit, which may be a lot.

I forgot the justification of why we need to keep track of the specific git commit.

If we don't really need the values for each git commit, I seem to recall that the last thing we proposed here was to report a histogram (without the `commit` label) in the `oc_apply` method. The value would be `current_time - git_commit_time`. The `git_commit_time` is not yet queriable via qontract-server, but it would be. So everytime you set up and initialize a new `sha` url for the server, you would also fetch the `git_commit_time` associated to it.

BTW, I liked the namedtuple _replace method, pretty cool :)

**Comment by rporres** [comment]
This looks cool. For the purposes of what we want to achieve, most likely we will need to include in the same metric the deletes and the replaces in the same metric, don't you think @jmelis. The annotation name `track` would probably be too generic.

**Comment by apahim** [comment]
As of now, the decorated function name is a dimension to the metric.

**Comment by rporres** [comment]
> As of now, the decorated function name is a dimension to the metric.

Yeah, that was what I understood from the code

**Comment by apahim** [comment]
> most likely we will need to include in the same metric the deletes and the replaces in the same metric

so all the decorated functions will contribute to the same histogram?

**Comment by rporres** [comment]
Forget about it, we can just sum with a regex and that should be it :)

File: utils/oc.py [hunk 1]
```diff
@@ -158,6 +159,7 @@ def remove_last_applied_configuration(self, namespace, kind, name):
                'kubectl.kubernetes.io/last-applied-configuration-']
         self._run(cmd)
 
+    @track
```
Comment (rporres) [inline utils/oc.py:162]: Please add tracking to `replace` and `delete` too

File: utils/metrics.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+run_time = Gauge(name='qontract_reconcile_last_run_seconds',
+                 documentation='Last run duration in seconds',
+                 labelnames=['integration', 'shards', 'shard_id'])
+
+run_status = Counter(name='qontract_reconcile_run_status',
+                     documentation='Status of the runs',
+                     labelnames=['integration', 'status',
+                                 'shards', 'shard_id'])
+
+func_runtime = Histogram(name='qontract_reconcile_function_runtime',
```
Comment (rporres) [inline utils/metrics.py:15]: would be `qontract_reconcile_function_elapsed_seconds_since_bundle_commit` too long? 😂

File: utils/metrics.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+                         documentation='Run time for tracked functions',
```
Comment (rporres) [inline utils/metrics.py:16]: We should specify the unit here

File: utils/metrics.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+                         labelnames=['name'],
+                         buckets=(.1, .5, 1.0, 5.0, 10.0, 50.0, 100.0, 500.0,
```
Comment (rporres) [inline utils/metrics.py:18]: What do you thing about this?
```suggestion
                         buckets=(10.0, 30.0, 60.0, 150.0, 300.0, 600.0, 1200.0, 1800.0,
```

File: utils/qrtracking.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import time
+
+from functools import wraps
+
+from reconcile.status import RunningState
+from utils.metrics import func_runtime
+
+
+def track(function):
```
Comment (rporres) [inline utils/qrtracking.py:9]: What about? Maybe too long (again)
```suggestion
def elapsed_seconds_from_commit_metric(function):
```

Comment (apahim) [reply utils/qrtracking.py:9]: That's terrible, but I like it :)

File: utils/metrics.py (hunk 4 above)
Comment (apahim) [reply utils/metrics.py:18]: sure +1

File: utils/metrics.py (hunk 2 above)
Comment (apahim) [reply utils/metrics.py:15]: not for me. hopefully also not for prometheus.

Comment (rporres) [reply utils/metrics.py:15]: It doesn't look like it : https://prometheus.io/docs/concepts/data_model/

**Comment by apahim** [comment]
@rporres @jmelis updated. The description shows how the metric is looking atm.

File: utils/oc.py [hunk 6]
```diff
@@ -128,6 +129,7 @@ def get_items(self, kind, **kwargs):
 
         return items
 
+    @elapsed_seconds_from_commit_metric
```
Comment (rporres) [inline utils/oc.py:132]: I don't think we need to track this ATM, only calls that write into k8s

**Review by rporres — APPROVED** [review]
Apart from a small details, looks great!

File: utils/oc.py (hunk 6 above)
Comment (apahim) [reply utils/oc.py:132]: sorry, leftover from testing :]

Comment (rporres) [reply utils/oc.py]: thx!

**Pull request merged**